### DISCUSSION TITLE
Performance optimizations

### DISCUSSION THREAD

**Discussion by Blebowski** [nickg/nvc#758] [General]
Hi @nickg , I have noticed you have optimized many things, including IEEE std_logic operations.
Profiling the latest master vs 1.10 release gives me about 20-23 % simulation speedup.

I was wondering whether you plan to do any further optimizations, or whether you are aware
of any other potential sweet-spots for performance optimizations.

**Comment by nickg** [comment]
The recent improvements came from profiling a design a user shared with with me privately. If anyone has large designs either public or private that they can share with me I'm happy to take a look. I've got some ideas for improving how drivers are represented that would reduce overhead, but most concrete performance improvements come from fixing an issue found in some particular benchmark.

**Comment by Blebowski** [comment]
OK, I see thanks. Do you think it would be worth-it to expose the profiler of the generated code to the user ? So that user can see which code takes the longest to simulate?

**Reply by nickg** [reply]
Maybe, but it would be a lot of work to make a profiler that's suitable for end-users and works across all platforms (esp. Windows and Mac). And I'm not sure how useful it would really be as most of the time is spent in the simulator runtime code rather than the user's code. For example from the NEORV32 top-level testbench:

```
# Overhead  Shared Object                     
# ........  ..................................
#
    66.27%  nvc                               
    23.18%  _NEORV32.NEORV32_TB_SIMPLE.elab.so
     7.83%  preload08.so                      
     2.43%  libc.so.6                         
     0.18%  [kernel.kallsyms]                 
     0.07%  ld-linux-x86-64.so.2              
     0.04%  libzstd.so.1.5.5                  
     0.00%  libLLVM-15.so.1       
```

So 66% inside the simulator itself and only 23% executing the user's code.

**Reply by Blebowski** [reply]
I see. You are right that its not that useful for user since most of the time is spent by the simulator itself.

**Reply by avelure** [reply]
Could it be that the NEORV32 is already rather optimized?
I usually run the Modelsim profiler once in a while to investigate bottlenecks and I can usually find one or two things to optimize, but once those things are fixed, it ends up being mostly simulator core or memory copy operations that are the primary hoggs.
Would it be possible to link back more of the calls in the simulator to actions in the user code?
